tools/uart-flash-server/ufserver.c

```diff
@@ -0,0 +1,418 @@
+#define CMD_HDR_ERASE 0x03
+#define CMD_ACK       0x06
+
+#define FIRMWARE_PARTITION_SIZE 0x4000
+#define SWAP_SIZE 0x4000
+#define UART_BITRATE 460800
+
+const char msgSha[]         = "Verifying SHA digest...";
+const char msgReadUpdate[]  = "Fetching update blocks ";
+const char msgReadSwap[]    = "Reading SWAP blocks    ";
+const char msgWriteUpdate[] = "Writing backup blocks  ";
+const char msgWriteSwap[]   = "Writing SWAP blocks    ";
+const char msgEraseUpdate[] = "Erase update blocks    ";
+const char msgEraseSwap[]   = "Erase swap blocks      ";
+
+extern uint16_t wolfBoot_find_header(uint8_t *haystack, uint16_t type, uint8_t **ptr);
+
+const char blinker[]="-\\|/";
+
+void printmsg(const char *msg)
+{
+    static const char *cur_msg = NULL;
+    static int b_idx = 0;
+    if (cur_msg != msg) {
+        cur_msg = msg;
+    }
+    b_idx++;
+    if (b_idx > 3)
+        b_idx = 0;
+    printf("\r[%c] %s\t\t\t", blinker[b_idx], msg);
+    fflush(stdout);
+}
+
+
+int hal_flash_write(uint32_t address, const uint8_t *data, int len)
+{
+    (void)address;
+    (void)data;
+    (void)len;
+    return -1;
+}
+
+int hal_flash_erase(uint32_t address, int len)
+{
+    (void)address;
+    (void)len;
+    return -1;
+}
+void hal_flash_unlock(void)
+{
+}
+void hal_flash_lock(void)
+{
+}
+
+#include <termio.h>
+#include <fcntl.h>
+#include <err.h>
+#include <linux/serial.h>
+
+static int rate_to_constant(int baudrate) {
+#define B(x) case x: return B##x
+    switch(baudrate) {
+        B(50);     B(75);     B(110);    B(134);    B(150);
+        B(200);    B(300);    B(600);    B(1200);   B(1800);
+        B(2400);   B(4800);   B(9600);   B(19200);  B(38400);
+        B(57600);  B(115200); B(230400); B(460800); B(500000);
+        B(576000); B(921600); B(1000000);B(1152000);B(1500000);
+default: return 0;
+}
+#undef B
+}
+
+/* Open serial port in raw mode, with custom baudrate if necessary */
+static int serial_open(const char *device, int rate)
+{
+	struct termios options;
+	struct serial_struct serinfo;
+	int fd;
+	int speed = 0;
+
+	/* Open and configure serial port */
+	if ((fd = open(device,O_RDWR|O_NOCTTY)) == -1)
+		return -1;
+
+	speed = rate_to_constant(rate);
+
+	if (speed == 0) {
+		/* Custom divisor */
+		serinfo.reserved_char[0] = 0;
+		if (ioctl(fd, TIOCGSERIAL, &serinfo) < 0)
+			return -1;
+		serinfo.flags &= ~ASYNC_SPD_MASK;
+		serinfo.flags |= ASYNC_SPD_CUST;
+		serinfo.custom_divisor = (serinfo.baud_base + (rate / 2)) / rate;
+		if (serinfo.custom_divisor < 1)
+			serinfo.custom_divisor = 1;
+		if (ioctl(fd, TIOCSSERIAL, &serinfo) < 0)
+			return -1;
+		if (ioctl(fd, TIOCGSERIAL, &serinfo) < 0)
+			return -1;
+		if (serinfo.custom_divisor * rate != serinfo.baud_base) {
+			warnx("actual baudrate is %d / %d = %f",
+			      serinfo.baud_base, serinfo.custom_divisor,
+			      (float)serinfo.baud_base / serinfo.custom_divisor);
+		}
+	}
+
+	fcntl(fd, F_SETFL, 0);
+	tcgetattr(fd, &options);
+	cfsetispeed(&options, speed ?: B38400);
```

Comment:
Does it make sense to use B460800, since that's what was used in the wolfBoot UART code?

Reply:
OK, default changed